src/lib.rs

```diff
@@ -1,31 +1,54 @@
+macro_rules! circular_increment {
+    ($value:expr, $max:expr) => {
+        $value = ($value + 1) % $max;
+    };
+}
 
 pub struct RingBuffer<T: Copy> {
     data: Vec<Option<T>>,
+    write_head : usize,
+    read_head : usize
 }
 
 impl <T : Copy> RingBuffer<T> {
     pub fn new(ring_size: usize) -> Self {
         RingBuffer {
             data: vec![None; ring_size],    // ring_size x None
+            write_head : 0,
+            read_head : 0
         }
     }
 
     pub fn number_of_elements(&self) -> usize {
-        0
+        self.data.iter().filter(|&item| item.is_some()).count()
     }
 
     /// adds a new element to the ring,
     /// if the ring is full, the oldest element is woverwritten
     pub fn add(&mut self, element: T) {
+        let element_count = self.data.len();
+        if self.number_of_elements() == self.data.len(){//buffer full, cycle
+            self.data[self.write_head] = Some(element);
```

Comment:
do you think we could somehow get rid of this 2 line duplication?

Reply:
Which 2 line duplication do you mean? L32 and L33? If yes, i already thought about making a generic macro for 1 to n circular increments with the same max value, but i didnt got it work

Reply:
maybe I'm missing a detail, but it looks to me (although it's late already - and I'm getting old XD) as if the lines 
```rust
self.data[self.write_head] = Some(element);
circular_increment!(self.write_head, element_count);
```
are in both blocks, the if and the else. So it looks like, that we write, and move the write pointer anyways, no matter whether we`re full or not